# Homework 5 - Logistic Regression

## Your Name Here

Student Name: Erika Noda

Student UT EID: edn427 

---

Partner Name:

Partner UT EID:

---

Date Created: 2/20/24

Date Last Modified:

---

Total Points 20. 

In [1]:
# Standard Headers
# You are welcome to add additional headers here if you wish
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Enable inline mode for matplotlib so that Jupyter displays graphs
%matplotlib inline

C:\Users\erika\AppData\Local\Temp\ipykernel_11732\572230382.py:3: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


## Heart Dataset 

In this Assignment we will work with some patients dataset. 

We have access to 303 patients data. The features are listed below. 

In [2]:
# Your code here
heart_df = pd.read_csv("Heart.csv")
heart_df

Age  Sex     ChestPain  RestBP  Chol  Fbs  RestECG  MaxHR  ExAng  \
0     63    1       typical     145   233    1        2    150      0   
1     67    1  asymptomatic     160   286    0        2    108      1   
2     67    1  asymptomatic     120   229    0        2    129      1   
3     37    1    nonanginal     130   250    0        0    187      0   
4     41    0    nontypical     130   204    0        2    172      0   
..   ...  ...           ...     ...   ...  ...      ...    ...    ...   
298   45    1       typical     110   264    0        0    132      0   
299   68    1  asymptomatic     144   193    1        0    141      0   
300   57    1  asymptomatic     130   131    0        0    115      1   
301   57    0    nontypical     130   236    0        2    174      0   
302   38    1    nonanginal     138   175    0        0    173      0   

     Oldpeak  Slope   Ca        Thal Target  
0        2.3      3  0.0       fixed     No  
1        1.5      2  3.0      normal    Yes  
2        2.6      2  2.0  reversable    Yes  
3        3.5      3  0.0      normal     No  
4        1.4      1  0.0      normal     No  
..       ...    ...  ...         ...    ...  
298      1.2      2  0.0  reversable    Yes  
299      3.4      2  2.0  reversable    Yes  
300      1.2      2  1.0  reversable    Yes  
301      0.0      2  1.0      normal    Yes  
302      0.0      1  NaN      normal     No  

[303 rows x 14 columns]

**Age:** The person’s age in years

**Sex:** The person’s sex (1 = male, 0 = female)

**ChestPain:** chest pain type

* Value 0: asymptomatic
* Value 1: atypical angina
* Value 2: non-anginal pain
* Value 3: typical angina

**RestBP:** The person’s resting blood pressure (mm Hg on admission to the hospital)

**Chol:** The person’s cholesterol measurement in mg/dl

**Fbs:** The person’s fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)
restecg: resting electrocardiographic results

* Value 0: showing probable or definite left ventricular hypertrophy by Estes’ criteria
* Value 1: normal
* Value 2: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)

**RestECG:** The person’s maximum heart rate achieved

**MaxHR:** Exercise induced angina (1 = yes; 0 = no)

**Oldpeak:** ST depression induced by exercise relative to rest (‘ST’ relates to positions on the ECG plot. See more here)

**Slope:** the slope of the peak exercise ST segment — 0: downsloping; 1: flat; 2: upsloping

* 0: downsloping; 
* 1: flat; 
* 2: upsloping

**Ca:** The number of major vessels (0–3)

**Thal:** A blood disorder called thalassemia Value 0: NULL (dropped from the dataset previously

* Value 1: fixed defect (no blood flow in some part of the heart)
* Value 2: normal blood flow
* Value 3: reversible defect (a blood flow is observed but it is not normal)

**Target:** Heart disease (1 = no, 0= yes)

# Task - 1 (5 points)
We want to use logistic regerssion to perdict if the patients will have heart problems or not. The column "Target" in our datasets includes data about heart diseases. If the patient had heart disease we have a 1 and if not a zero. 

Prepare your data set for predicting heart disease ("Target" column) out of 3 features:

* Age of the patient (Column **"Age"**)
* Gender of the patient (male or female - Column **"Sex"**)
* Cholestrol level of the patient (Column **"Chol"**) 


Split your data into 80% traning data and 20% test data. 

In [30]:
# Add your code Here! 
# create smaller data set tht has columns: target, age, sex, and chol
df = heart_df.copy()
df = df[['Target', 'Age', 'Sex', 'Chol']]
df

Target  Age  Sex  Chol
0       No   63    1   233
1      Yes   67    1   286
2      Yes   67    1   229
3       No   37    1   250
4       No   41    0   204
..     ...  ...  ...   ...
298    Yes   45    1   264
299    Yes   68    1   193
300    Yes   57    1   131
301    Yes   57    0   236
302     No   38    1   175

[303 rows x 4 columns]

In [31]:
# check nulls
df.isnull().sum()

Target    0
Age       0
Sex       0
Chol      0
dtype: int64

In [32]:
# replace "No" and "Yes" in Target column with 0 and 1 based off of information given
df['Target'] = df['Target'].map({'No': 1, 'Yes': 0})
df

Target  Age  Sex  Chol
0         1   63    1   233
1         0   67    1   286
2         0   67    1   229
3         1   37    1   250
4         1   41    0   204
..      ...  ...  ...   ...
298       0   45    1   264
299       0   68    1   193
300       0   57    1   131
301       0   57    0   236
302       1   38    1   175

[303 rows x 4 columns]

In [33]:
# summary stats- make sure they make sense
# min for Sex should be 0 and max should be 1
# min for Target should be 0 and max should be 1
df.describe()

Target         Age         Sex        Chol
count  303.000000  303.000000  303.000000  303.000000
mean     0.541254   54.438944    0.679868  246.693069
std      0.499120    9.038662    0.467299   51.776918
min      0.000000   29.000000    0.000000  126.000000
25%      0.000000   48.000000    0.000000  211.000000
50%      1.000000   56.000000    1.000000  241.000000
75%      1.000000   61.000000    1.000000  275.000000
max      1.000000   77.000000    1.000000  564.000000

In [34]:
# double check that the correlation between the age, sex, and chol are not too high
df.corr()

Target       Age       Sex      Chol
Target  1.000000 -0.223120 -0.276816 -0.085164
Age    -0.223120  1.000000 -0.097542  0.208950
Sex    -0.276816 -0.097542  1.000000 -0.199915
Chol   -0.085164  0.208950 -0.199915  1.000000

In [44]:
# split data into test group and training group
from sklearn.model_selection import train_test_split

y = df['Target']
X = df.drop('Target', axis=1)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=123)

In [45]:
# view training and test data
X_train

Age  Sex  Chol
102   57    0   303
261   58    0   319
228   54    1   206
288   56    1   221
78    48    1   245
..   ...  ...   ...
106   59    1   177
83    68    1   274
17    54    1   239
230   52    0   196
98    52    1   201

[242 rows x 3 columns]

In [46]:
y_train

102    1
261    0
228    0
288    1
78     1
      ..
106    0
83     0
17     1
230    1
98     1
Name: Target, Length: 242, dtype: int64

In [47]:
X_test

Age  Sex  Chol
11    56    0   294
292   44    1   169
269   42    1   180
268   40    1   223
94    63    0   252
..   ...  ...   ...
95    52    1   255
26    58    0   340
259   57    1   261
143   64    1   309
42    71    0   302

[61 rows x 3 columns]

In [49]:
y_test


11     1
292    0
269    1
268    0
94     1
      ..
95     0
26     1
259    0
143    0
42     1
Name: Target, Length: 61, dtype: int64

# Task 2 - (10 points)

Generate a logistic regression model using your training data. 

Print out the accuracy if your model. 

In [58]:
# Add your code Here! 
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# TODO: check if fit_intercept is true or false... I think it should be false
model = LogisticRegression(fit_intercept=False)
model.fit(X_train, y_train)

y_pred = model.predict(X_test)
print('model accuracy:', accuracy_score(y_test,y_pred))

model accuracy: 0.6229508196721312


# Task 3 - (5 points)


Generate the classification report for your Logistic regresion model and interpret your results regarding precision, recall and f1-score. 


In [59]:
# Add your code Here! 
from sklearn.metrics import classification_report

print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.56      0.82      0.67        28
           1       0.75      0.45      0.57        33

    accuracy                           0.62        61
   macro avg       0.66      0.64      0.62        61
weighted avg       0.66      0.62      0.61        61



In [ ]:
# TODO: interpret the results

# Task 4 - Optional Task (2 extra points)


Which other feature of your data can you use to improve your prediction? Build other models and improve your prediction. 
